cmd/deployment-check/main.go

```diff
@@ -0,0 +1,259 @@
+	defaultCheckContainerPort = int32(80)
+
+	// Default load balancer port used for check.
+	defaultCheckLoadBalancerPort = int32(80)
+
+	// Default k8s manifest resource names.
+	defaultCheckDeploymentName = "kh-deployment-check-deployment"
+	defaultCheckServiceName    = "kh-deployment-check-service"
+
+	// Default namespace for the check to run in.
+	defaultCheckNamespace = "kuberhealthy"
+
+	// Default number of replicas the deployment should bring up.
+	defaultCheckDeploymentReplicas = 2
+
+	defaultCheckTimeLimit             = time.Duration(time.Minute * 5)
+	defaultShutdownGracePeriodSeconds = 30 // grace period for the check to shutdown after receiving a shutdown signal
+)
+
+func init() {
+
+	// Parse incoming debug settings.
+	parseDebugSettings()
+
+	// Parse all incoming input environment variables and crash if an error occurs
+	// during parsing process.
+	parseInputValues()
+
+	// Allocate channels.
+	signalChan = make(chan os.Signal, 3)
+	doneChan = make(chan bool)
+}
+
+func main() {
+
+	// Create a timestamp reference for the deployment;
+	// also to reference against deployments that should be cleaned up.
+	now = time.Now()
+	log.Debugln("Allowing this check", checkTimeLimit, "to finish.")
+	ctx, ctxCancel = context.WithTimeout(context.Background(), checkTimeLimit)
+
+	// Create a kubernetes client.
+	var err error
+	client, err = kubeClient.Create(kubeConfigFile)
+	if err != nil {
+		errorMessage := "failed to create a kubernetes client with error: " + err.Error()
+		reportErrorsToKuberhealthy([]string{errorMessage})
+		return
+	}
+	log.Infoln("Kubernetes client created.")
+
+	// Start listening to interrupts.
+	go listenForInterrupts()
+
+	// Start deployment check.
+	err = runDeploymentCheck()
+	if err != nil {
+		log.Errorln("Error when running check:", err)
+		os.Exit(1)
+	}
+}
+
+// listenForInterrupts watches the signal and done channels for termination.
+func listenForInterrupts() {
+
+	// Relay incoming OS interrupt signals to the signalChan.
+	signal.Notify(signalChan, os.Interrupt, os.Kill)
+	<-signalChan // This is a blocking operation -- the routine will stop here until there is something sent down the channel.
+	log.Infoln("Received an interrupt signal from the signal channel.")
+
+	log.Debugln("Cancelling context.")
+	ctxCancel() // Causes all functions within the check to return without error and abort. NOT an error
+	// condition; this is a response to an external shutdown signal.
+
+	// Clean up pods here.
+	log.Infoln("Shutting down.")
+
+	select {
+	case <-signalChan:
+		// If there is an interrupt signal, interrupt the run.
+		log.Warnln("Received a secsond interrupt signal from the signal channel.")
+	case err := <-cleanUpAndWait():
+		// If the clean up is complete, exit.
+		log.Infoln("Received a complete signal, clean up completed.")
+		if err != nil {
+			log.Errorln("failed to clean up check resources properly:", err.Error())
+		}
+	case <-time.After(time.Duration(shutdownGracePeriodSeconds)):
```

Comment:
make sure that `shutdownGracePeriodSeconds` always aligns with the `khcheck` pod spec.  You can define this variable in there and it defaults to 30s.